backend/src/order.ts

```diff
@@ -354,162 +310,173 @@ export function getAllForShow(show_id: number): Promise<IAdminOrderListItem[]> {
+
+    log.info('Created payment for order', {order_id});
+    return resp;
+  } catch (err) {
+    log.error('Failed to prepare payment', {error: err, order_id: order_id});
+    throw err;
   }
 }
 
-export function paymentDone(order_id: number, params): Promise<IOrder> {
-  log.info('Payment done - verifying', {order_id: order_id});
+export async function paymentDone(order_id: number, req: Request): Promise<any> {
+  try {
+    log.log('debug', 'Create payment handler', {order_id});
+    const handler = await getPaymentHandler(order_id);
 
-  var verification = [PAYTRAIL_PREFIX + order_id, params.TIMESTAMP, params.PAID, params.METHOD, config.paytrail.password].join('|');
-  var verification_hash = md5(verification).toUpperCase();
+    log.log('debug', 'Verify success', {order_id});
 
-  if (verification_hash !== params.RETURN_AUTHCODE) {
-    log.error('Hash verification failed!',
-      {order_id: order_id, verification_hash: verification_hash, return_authcode: params.RETURN_AUTHCODE});
-    throw 'Hash verification failed!';
+    const resp = await handler.verifySuccess(req);
+    log.info('Payment success verification succeeded', {order_id});
+
+    if ( await updatePaymentStatusToPaid(order_id) ) {
+      log.log('debug', 'Sending tickets', {order_id});
+      await sendTickets(order_id);
+    }
+
+    log.log('debug', 'Getting updated order', {order_id});
+    const order = await get(order_id);
+    return order;
+  } catch (err) {
+    log.error('Payment success verification failed', {error: err, order_id});
+    throw err;
   }
+}
 
-  log.info('Verification ok', {order_id: order_id});
+async function updatePaymentStatusToPaid(order_id: number): Promise<boolean> {
+  let conn: IConnection;
+  
+  try {
+    conn = await db.beginTransaction();
+  } catch (err) {
+    log.error('Payment status update to paid failed', {error: err, order_id});
+    throw err;
+  }
 
-  return db.query('select status from nk2_orders where id = :order_id', {order_id: order_id})
-  .then((rows) => {
-    if (rows[0].status === 'paid') {
+  try {
+    const res = await db.query('select status from nk2_orders where id = :order_id', {order_id}, conn);
+    if(res[0].status === 'paid') {
       log.info('Order was already paid', {order_id: order_id});
-      return null;
+      await db.rollback(conn);
+      return false;
     }
-    return db.query('update nk2_orders set status = "paid", payment_id = :payment_id where id = :order_id', { order_id: order_id, payment_id: params.PAID})
-    .then(() => sendTickets(order_id))
-    .catch((err) => {
-      log.error('Updating payment status failed', {error: err, order_id: order_id});
-      return Promise.reject('Updating payment status failed');
-    });
-  })
-  .then(() => {
-    return get(order_id);
-  });
 
+    log.log('debug', 'Updating order payment status', {order_id});
+    await db.query('update nk2_orders set status = "paid" where id = :order_id',{order_id}, conn);
```

Comment:
I think UPDATE returns the number of changed rows, so we could just use it to check if anything was modified instead of the TRX->SELECT->UPDATE->COMMIT.  Not necessary to change now.

Reply:
Yeah, let's seet this at some other point.